core/dnsserver/server_quic.go

```diff
@@ -0,0 +1,287 @@
+func (s *ServerQUIC) serveQUICConnection(conn quic.Connection) {
+	for {
+		// In DoQ, one query consumes one stream.
+		// The client MUST select the next available client-initiated bidirectional
+		// stream for each subsequent query on a QUIC connection.
+		stream, err := conn.AcceptStream(context.Background())
+		if err != nil {
+			closeQUICConn(conn, DoQCodeNoError)
+			return
+		}
+
+		go func() {
+			s.serveQUICStream(stream, conn)
+
+			// The server MUST send the response(s) on the same stream and MUST
+			// indicate, after the last response, through the STREAM FIN
+			// mechanism that no further data will be sent on that stream.
+			_ = stream.Close()
+		}()
+
+	}
+}
+
+func (s *ServerQUIC) serveQUICStream(stream quic.Stream, conn quic.Connection) {
+	bufPtr := s.bytesPool.Get().(*[]byte)
+	defer s.bytesPool.Put(bufPtr)
+
+	buf := *bufPtr
+	_, err := readAll(stream, buf)
+
+	// io.EOF does not really mean that there's any error, it is just
+	// the STREAM FIN indicating that there will be no data to read
+	// anymore from this stream.
+	if err != nil && err != io.EOF {
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	req := &dns.Msg{}
+
+	// Drafts of the RFC9250 did not require the 2-byte prefixed message length.
+	// Thus, we are only supporting the official version (DoQ v1).
+	err = req.Unpack(buf[2:])
+	if err != nil {
+		clog.Debug("unpacking quic packet: %s", err)
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	if !validRequest(req) {
+		// If a peer encounters such an error condition, it is considered a
+		// fatal error. It SHOULD forcibly abort the connection using QUIC's
+		// CONNECTION_CLOSE mechanism and SHOULD use the DoQ error code
+		// DOQ_PROTOCOL_ERROR.
+		closeQUICConn(conn, DoQCodeProtocolError)
+
+		return
+	}
+
+	w := &DoQWriter{
+		localAddr:  conn.LocalAddr(),
+		remoteAddr: conn.RemoteAddr(),
+		stream:     stream,
+		Msg:        req,
+	}
+
+	dnsCtx := context.WithValue(context.Background(), Key{}, s.Server)
+	dnsCtx = context.WithValue(dnsCtx, LoopKey{}, 0)
+	s.ServeDNS(dnsCtx, w, req)
+}
+
+// ListenPacket implements caddy.UDPServer interface.
+func (s *ServerQUIC) ListenPacket() (net.PacketConn, error) {
+	p, err := reuseport.ListenPacket("udp", s.Addr[len(transport.QUIC+"://"):])
+	if err != nil {
+		return nil, err
+	}
+
+	s.quicListener, err = quic.Listen(p, s.tlsConfig, s.quicConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// OnStartupComplete lists the sites served by this server
+// and any relevant information, assuming Quiet is false.
+func (s *ServerQUIC) OnStartupComplete() {
+	if Quiet {
+		return
+	}
+
+	out := startUpZones(transport.QUIC+"://", s.Addr, s.zones)
+	if out != "" {
+		fmt.Print(out)
+	}
+}
+
+// Stop stops the server. It blocks until the server is totally stopped.
+func (s *ServerQUIC) Stop() error { return nil }
+
+// Shutdown stops the server (non gracefully).
+func (s *ServerQUIC) Shutdown() error { return nil }
```

Comment:
shoudn't you close the listener here?
```suggestion
func (s *ServerQUIC) Shutdown() error { 
	if s.quicListener != nil {
		return s.quicListener.Close()
	}
	return nil
}
```

Reply:
Did it, thanks!